### DISCUSSION TITLE
Lua script database table

### DISCUSSION THREAD

**Discussion by meleu** [RetroAchievements/RAWeb#733] [Ideas]
@luchaos 

This is the first approach I had in mind for a Lua script database table:

**[EDIT: at a second thought the concept below isn't enough. Let's discuss a better approach later]**

**LuaScript**

- `ID` - int(10) unsigned
- `GameID` - int(10) unsigned
- `Author` - varchar(32)
- `DateCreated` - timestamp
- `DateModified` - timestamp
- `ModifiedBy` - varchar(32)
- `Script` - ?? (textfile)

(I'm not sure how pure textfiles are stored in the database. Is it just a varchar with high limit?)

**NOTE**: As far as I understood, there is not a big concern to design a definitive implementation. The RAIntegration devs (@Jamiras and @leiradel) just need some kind of interaction with the server for the first development tests. **We don't plan to release this feature before RAWeb 2.**

They will need something like `uploadlua` in `dorequest.php` to upload Lua scripts, and maybe we should make `patch` return the Lua script amongst the other data.

But if you take care of the database table creation, I can code the requests.

**Comment by leiradel** [comment]
Do we need the author field? What happens if another person edits the script to add new functions for new or improved cheevos?

**Comment by meleu** [comment]
> What happens if another person edits the script to add new functions for new or improved cheevos?

The `ModifiedBy` field will be updated.

**Comment by GameDragon2k** [comment]
@leiradel 
Then the `ModifiedBy` gets changed. The `Author` field is the original founding author.

**Comment by leiradel** [comment]
What if a third person modifies the script? Wouldn't it be better to just add you nick as a comment in the code?

**Comment by GameDragon2k** [comment]
@leiradel `ModifiedBy` is the last person who edited.
`Author` is the person who created it.

**Comment by leiradel** [comment]
Sure, I just don't get why save this info. It doesn't tell the history of changes.

I'm probably just missing something though.

**Comment by luchaos** [comment]
history will be stored independently from that, including the contents (to show diffs) - the single entry can be understood as latest snapshot then. the column could be named `updated_by` as well. but those changes will be made in v2 anyways.
~~what i don't understand, why don't we just use the same table as for the curremt memory notation triggers?
my plan was to rename MemAddress to simply `trigger`, and have another `trigger_type` which would be either `ramem` or `lua`~~
(just realized those lua scripts will be assigned to games as a whole)

**Comment by meleu** [comment]
The design of that table I proposed in the OP was highly influenced by the current existing tables. I didn't spend much thinking on it because having a definitive solution is not that important. It's just for tests.

When RAWeb V2 become a reality we can spend more time thinking on a better design.

**Comment by leiradel** [comment]
How is this going to work from the clients perspective? Right now **rcheevos** assumes all Lua functions are implemented in one script file, and regular cheevos use a new Lua operand type in the `MemAddr`, which starts with `@`.

This is incompatible with having a different kind of cheevo in the database for Lua cheevos.

**Comment by meleu** [comment]
If I understood correctly, the way **rcheevos** is managing Lua scripts for cheevos is in accordance with what we discussed on discord, and I think that's fine.

The implementation I envisage (abstracting many details):
- a Lua script is bound to an individual game ID.
- when the client does a `dorequest.php?r=patch`, the server gives the same thing as it gives currently plus the respective game's Lua script (if it exist).
- the client keeps the script locally along with the other cheevos data.

@leiradel is it enough for your question or do we need to discuss more details?

(by the way, maybe @Jamiras can also share some thoughts here)

**Comment by leiradel** [comment]
Will the two files, the JSON and the Lua script, be somehow packed, say ZIP, to be returned by the same request? I'd rather have the Lua script embedded in the JSON, or two separate requests i.e. `dorequest.php?r=script`.

**Comment by meleu** [comment]
I'm thinking in something like this: 

```json
{
  "Success": true,
  "PatchData": {
    "ID": 1,
    "Title": "Sonic the Hedgehog",
    "ConsoleID": 1,
    "ForumTopicID": 112,
    "Flags": 0,
    "ImageIcon": "/Images/016743.png",
    "ImageTitle": "/Images/000009.png",
    "ImageIngame": "/Images/000010.png",
    "ImageBoxArt": "/Images/016727.png",
    "Publisher": "SEGA®",
    "Developer": "SEGA®",
    "Genre": "Platformer",
    "Released": "June 23, 1991",
    "IsFinal": false,
    "ConsoleName": "Mega Drive",
    "RichPresencePatch": "...",

    "LuaScript": "===== L U A   S C R I P T   C O N T E N T =====",

    "Achievements": [
      {
        "ID": "..."
      }
    ],
    "Leaderboards": [
      {
        "ID": "..."
      }
    ]
  }
}
```

what do you think?

**Comment by Jamiras** [comment]
I was assuming something along those lines as well (returned inside the PatchData).
